src/System.Private.ServiceModel/tools/PRService/sync.ashx

```diff
@@ -0,0 +1,227 @@
+        {
+            context.Response.StatusCode = 400;
+        }
+    }
+
+    // Check prerequisites for whether or not we can run
+    private bool CheckPrerequisites(StringBuilder executionResult)
+    {
+        if (!File.Exists(_gitExecutablePath))
+        {
+            executionResult.AppendFormat("Could not find or insufficient permissions to access the git executable at path '{0}'", _gitExecutablePath);
+            return false;
+        }
+
+        if (!Directory.Exists(_gitRepoPath))
+        {
+            executionResult.AppendFormat("Could not find or insufficient permissions to access the git repo at path '{0}'", _gitRepoPath);
+            return false;
+        }
+
+        // Do a sanity execute of the git executable 
+
+        try
+        {
+            string[] gitCommands = new string[]
+            {
+                "status"
+            };
+
+            RunGitCommands(gitCommands, executionResult);
+        }
+        catch (Exception ex)
+        {
+            executionResult.AppendFormat("Exception caught while attempting to start '{0}'", _gitExecutablePath);
+            executionResult.Append(ex.ToString());
+            return false;
+        }
+
+        return true;
+    }
+
+    private bool SyncToPr(string pr, StringBuilder executionResult)
+    {
+        string[] gitCommands = new string[]
+        {
+            "clean -fdx",
+            "checkout master",
+            "fetch --all --prune",
+            "merge --ff-only origin/master",
+            string.Format("fetch origin pull/{0}/head:pr/{0}", pr),
+            string.Format("checkout pr/{0}", pr)
+        };
+
+        return RunGitCommands(gitCommands, executionResult);
+    }
+
+    private bool SyncToSha(string sha, StringBuilder executionResult)
+    {
+        string[] gitCommands = new string[]
+        {
+            "clean -fdx",
+            "checkout master",
+            "fetch --all --prune",
+            "merge --ff-only origin/master",
+            string.Format("checkout {0}", sha)
+        };
+
+        return RunGitCommands(gitCommands, executionResult);
+    }
+
+    private void CleanupMergedBranches(StringBuilder executionResult)
+    {
+        RunGitCommands(new string[] { "checkout master" }, executionResult);
+
+        StringBuilder branches = new StringBuilder();
+        RunGitCommands(new string[] { "branch" }, branches);
+
+        var branchesList =
+            branches.ToString().Split(new char[] { '\n', '\r' }, StringSplitOptions.RemoveEmptyEntries);
+
+        List<string> commands = new List<string>();
+
+        for (int i = 0; i < branchesList.Length; i++)
+        {
+            // Don't try to delete "master" branch, since we're checked out to it
+            if (!branchesList[i].EndsWith("master"))
+            {
+                commands.Add(string.Format("branch -D {0}", branchesList[i]));
+            }
+        }
+
+        RunGitCommands(commands.ToArray(), executionResult);
+    }
```

Comment:
It's necessary for repo maintenance. 
I anticipate that it will be called either on every sync or as a scheduled task. Thoughts on where it should be done?